### PR TITLE
[Merged by Bors] - feat: port Data.Sum.Basic

### PR DESCRIPTION
**Pull request by pechersky**
Changed the import from `Data.Option.Basic` to `Logic.Function.Basic`
because that is enough

### CODE REVIEW

File: Mathlib/Data/Sum/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+`α ⊕ β` is the type made of a copy of `α` and a copy of `β`. It is also called *disjoint union*.
+
+## Main declarations
+
+* `Sum.get_left`: Retrieves the left content of `x : α ⊕ β` or returns `none` if it's coming from
+  the right.
+* `Sum.get_right`: Retrieves the right content of `x : α ⊕ β` or returns `none` if it's coming from
+  the left.
+* `Sum.is_left`: Returns whether `x : α ⊕ β` comes from the left component or not.
+* `Sum.is_right`: Returns whether `x : α ⊕ β` comes from the right component or not.
```
Comment (semorrison) [inline Mathlib/Data/Sum/Basic.lean:22]:
```suggestion
* `Sum.getLeft`: Retrieves the left content of `x : α ⊕ β` or returns `none` if it's coming from
  the right.
* `Sum.getRight`: Retrieves the right content of `x : α ⊕ β` or returns `none` if it's coming from
  the left.
* `Sum.isLeft`: Returns whether `x : α ⊕ β` comes from the left component or not.
* `Sum.isRight`: Returns whether `x : α ⊕ β` comes from the right component or not.
```

File: Mathlib/Data/Sum/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+  | inl _ => false
+  | inr _ => true
+
+variable {x y : Sum α β}
+
+theorem get_left_eq_none_iff : x.getLeft = none ↔ x.isRight := by
+  cases x <;> simp only [getLeft, isRight, eq_self_iff_true]
+
+theorem get_right_eq_none_iff : x.getRight = none ↔ x.isLeft := by
+  cases x <;> simp only [getRight, isLeft, eq_self_iff_true]
```
Comment (semorrison) [inline Mathlib/Data/Sum/Basic.lean:95]: Perhaps rename and #align these, to use `getLeft` and `getRight`.

File: Mathlib/Data/Sum/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+* `Sum.get_left`: Retrieves the left content of `x : α ⊕ β` or returns `none` if it's coming from
+  the right.
+* `Sum.get_right`: Retrieves the right content of `x : α ⊕ β` or returns `none` if it's coming from
+  the left.
+* `Sum.is_left`: Returns whether `x : α ⊕ β` comes from the left component or not.
+* `Sum.is_right`: Returns whether `x : α ⊕ β` comes from the right component or not.
+* `Sum.map`: Maps `α ⊕ β` to `γ ⊕ δ` component-wise.
+* `Sum.elim`: Nondependent eliminator/induction principle for `α ⊕ β`.
+* `Sum.swap`: Maps `α ⊕ β` to `β ⊕ α` by swapping components.
+* `Sum.lex`: Lexicographic order on `α ⊕ β` induced by a relation on `α` and a relation on `β`.
```
Comment (semorrison) [inline Mathlib/Data/Sum/Basic.lean:26]:
```suggestion
* `Sum.Lex`: Lexicographic order on `α ⊕ β` induced by a relation on `α` and a relation on `β`.
```

File: Mathlib/Data/Sum/Basic.lean (hunk 2 above)
Comment (semorrison) [reply Mathlib/Data/Sum/Basic.lean:95]: Actually, I'm uncertain on this sort of renaming, and have asked on [zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/naming.20conventions/near/305785765).

Comment (pechersky) [reply Mathlib/Data/Sum/Basic.lean]: Can you check how I've written it and help suggest how to make it correct, please?

File: Mathlib/Data/Sum/Basic.lean [hunk 4]
```diff
@@ -68,30 +69,38 @@ def getLeft : Sum α β → Option α
...
 def isRight : Sum α β → Bool
   | inl _ => false
   | inr _ => true
 
+#align sum.is_right Sum.isRight
+
 variable {x y : Sum α β}
 
-theorem get_left_eq_none_iff : x.getLeft = none ↔ x.isRight := by
+theorem getLeft_eq_none_iff : x.getLeft = none ↔ x.isRight := by
```
[the remark this answers is not in the document]
Comment (pechersky) [reply Mathlib/Data/Sum/Basic.lean:100]: Thanks, makes sense. Is there a way to get `mathport` to realize that the `getLeft` above is the same as the token mentioned in this lemma name?

[the remark this answers is not in the document]
Comment (digama0) [reply Mathlib/Data/Sum/Basic.lean:92]: Not currently. And in any case there would be way too many false positives for that kind of thing.

File: Mathlib/Data/Sum/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+
+end Lex
+
+end Sum
+
+open Sum
+
+namespace Function
+
+theorem Injective.sum_elim {f : α → γ} {g : β → γ} (hf : injective f) (hg : injective g)
```
Comment (semorrison) [inline Mathlib/Data/Sum/Basic.lean:418]:
```suggestion
theorem injective.sum_elim {f : α → γ} {g : β → γ} (hf : injective f) (hg : injective g)
```
Hopefully this won't require an #align once #498 is merged.

**Comment by semorrison** [comment]
bors d+

**Comment by semorrison** [comment]
bors d-

**Comment by semorrison** [comment]
Sorry, hold up here. Mario says we should capitalize `Injective`.

**Comment by pechersky** [comment]
Got it!

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**